### DISCUSSION TITLE
Copy-PnpFile - Copies containing folders of files.

### DISCUSSION THREAD

**Discussion by I-BhushanSalunkhe** [pnp/powershell#296] [Q&A]
### Reporting an Issue or Missing Feature

We have a requirement of copying files from one document library to another document library with different site collections. However, in source document library, we have files inside nested folders, but we want to copy all files not folders to the target document library.
When we are trying to copy files, containing folders are also getting copied to target library.

### Expected behavior 

Only files should get copied to the target library.

### Actual behavior 

When we are trying to copy files, containing folders are also getting copied to target library.

### Steps to reproduce behaviour

`Copy-PnPFile -SourceUrl "Source Document URL" -TargetUrl "Target Library URL" -Force;`

### What is the version of the Cmdlet module you are running?

> 3.29.2101.0
> SharePointPnPPowerShellOnline
> PSGallery
> Microsoft 365 Patterns and Practices PowerShell Cmdlets for SharePoint Online

### Which operating system/environment are you running PnP PowerShell on?

- [x] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by veronicageek** [comment]
Hi @Bhushan12986,
`Copy-PnPFile` will copy a file or a folder (see [documentation](https://github.com/pnp/powershell/blob/dev/documentation/Copy-PnPFile.md)). Have you tried any conditional approaches with maybe the `FileSystemObjectType`? 

I had no issue on my side copying just the files (no folders) when using this approach (_change SitePnP1 and SitePnP2 to your sites_):

```
$allItems = Get-PnPListItem -List "Shared Documents" -FolderServerRelativeUrl "/sites/SitePnP1/Shared Documents"

foreach ($item in $allItems) {
    if ($item.FileSystemObjectType -eq "File"){
        Write-Host "Copying file: $($item.FieldValues.FileLeafRef)"
        Copy-PnPFile -SourceUrl "$($item.FieldValues.FileRef)" -TargetUrl "/sites/SitePnP2/TargetLibrary" -Force
    }
}
```

![image](https://user-images.githubusercontent.com/25181757/107966436-c4615e80-6fa3-11eb-9e95-561f444031ad.png)

![image](https://user-images.githubusercontent.com/25181757/107966481-d3481100-6fa3-11eb-9171-58fed16601ec.png)

![image](https://user-images.githubusercontent.com/25181757/107966509-dcd17900-6fa3-11eb-8ccd-d93d1a3eb898.png)

**Comment by I-BhushanSalunkhe** [comment]
Hi @veronicageek 

We have exported source library details to a CSV file and we are reading data from it only as we need to update some metadata/properties which is also part of the same CSV file of copied files. We are using Path and Name fields' combination as source document URL as below, still containing folders are getting copied to the target document library.

`$csvSourceFile = Import-Csv $sourceCSVFile;
ForEach ($Row in $csvSourceFile) {
	Copy-PnPFile -SourceUrl "/$($Row.Path)/$($Row.Name)" -TargetUrl "$($destinationDocumentLibrary)" -Force;
}`

**Comment by veronicageek** [comment]
@Bhushan12986 - Are you explicitly giving each file name in your csv file? If not, please use the `FileSystemObjectType` to target only the files and not the folders.

**Comment by I-BhushanSalunkhe** [comment]
Hi @veronicageek,

Yes, we are reading each file name explicitly from CSV file. Path and Name are columns from CSV itself for file path and name respectively. CSV contains only details of the files, it does not contain any row of folder details, we have removed folder rows from it.

We also tried with below code but output is same, it is copying containing folders too.

```
#Get All Files from the document library - In batches of 500
Write-host -ForegroundColor Magenta "Getting All Documents from Source Document Library.";
$sourceDocuments = Get-PnPListItem -List $sourceLibrary.Title -PageSize 500 -Web $sourceWeb -Connection $SourceConn;

#Get the Data from CSV for reference
Write-host -ForegroundColor Yellow "Getting Records from CSV.";
$csvSourceFiles = Import-Csv $sourceCSVFile;

ForEach ($Row in $csvSourceFiles) {
  Write-host -ForegroundColor Magenta "Getting details of file name: $($Row.Path)/$($Row.Name)";
  $sourceDocument = $sourceDocuments | Where-Object { $_.ID -eq $Row.ID };
  
  Copy-PnPFile -SourceUrl "$($sourceDocument.FieldValues.FileRef)" -TargetUrl "$($destinationDocumentLibrary)" -Force;
}
```

Also, is it possible to return copied document using **Copy-PnPFile** to get the ID of the new copied document in order to update metadata/properties?

**Comment by veronicageek** [comment]
Moving into a discussion as this is more related to a script and not an issue with the cmdlet.

**Comment by veronicageek** [comment]
@Bhushan12986 - Can you post an example of your CSV file with dummy data? And also try on a subset with the condition using `FileSystemObjectType -eq "File"`

**Reply by I-BhushanSalunkhe** [reply]
@veronicageek,

Thank you for looking into this, will try with your suggestion and find the attached dummy data of CSV file.

[SampleDummyData.xlsx](https://github.com/pnp/powershell/files/6000481/SampleDummyData.xlsx)

**Comment by lsg-saco** [comment]
From my testing this is a bug with this version of pnp powershell that is fixed in the current version of PnP.Powershell. To work around it you simply need to add a slash ( + "/") to the end of the SourceUrl field.
